design/service-apis-implementation.md

```diff
@@ -0,0 +1,342 @@
+
+### Status management
+
+Each object in the Service APIs set has its own status definition. In general, Contour will update the status of any object that comes into scope with its details, ensuring that the `observedGeneration` field is set correctly.
+When objects fall out of scope, any status set by Contour will not be removed.
+It's expected that things that check the status will also check the `observedGeneration` to check if the status information is up-to-date with the current object generation.
+
+### Per-object design notes
+
+#### GatewayClass
+
+Contour does not watch or interact with GatewayClass in any way.
+
+#### Gateway
+Contour will watch its configured Gateway object.
+
+Contour is not able to specify the address that the Envoy deployment will listen on, and so will ignore any entries in the Gateway's `spec.addresses` field.
+Contour will not update the `status.addresses` field.
+
+For Contour, the key Gateway section are the Listeners. These define how an implementation should listen for traffic described by Routes.
+
+(The remainder of this document assumes that [#3263](https://github.com/projectcontour/contour/pull/3263) has been approved and merged.)
+
+For Listeners:
+- Listeners inside Gateways will be merged where possible.
+- Listeners that specify a port that the Contour controller considers the default secure port will have a HTTP->HTTPS redirect created for them.
+- Listeners that specify a port that is not exposed via the Envoy service will be rejected.
+- Conflicts within a Gateway will result in the relevant Listeners both being rejected (as there is no way to determine which one was first).
+- Listeners are considered mergeable if all the fields out of `hostname`, `port`, and `protocol` match, with some additional rules around TLS.
+- Further merging rules are specified in the detailed design below.
+- Listeners that refer to any other Route than HTTPRoute or TLSRoute will be ignored, and a condition placed on the corresponding `status.listeners[]` object saying that it was ignored because those objects are not supported.
+
+
+The Gateway may supply TLS config, in which case it is used as a default.
+The TLS config may be overridden by HTTPRoutes if the Gateway `spec.listeners[].tls.routeOverride` is set to `Allow`.
+This allows the Gateway to configure a default TLS certificate.
+Note that this feature has fiddly interactions with the various places in which a hostname may be specified;
+this will require careful test case design.
+
+The output of this watcher is both Envoy configuration, and a list of kind/name/namespace details for Routes to watch.
+
+Contour will only update the `status` of Gateway objects.
+
+### HTTPRoute
+Contour will watch all HTTPRoute objects, and filter for entries matching a label selector in its configured Gateway's `spec.listeners[].routes`,
```

Comment:
Did I understand it correctly that `HTTPRoute` and `TLSRoute` resources are first selected from Gateway's own namespace or from namespaces listed in Gateway's `spec.listeners[].routes.namespaces` (if given), and after that optional label selector `spec.listeners[].routes.selector` is applied (if given).  With the current wording this two level process might be bit unclear. 
 
Since the selection logic is more `Gateway` topic, it could be moved from HTTPRoute and TLSRoute chapters to Gateway chapter, but it is readable as it is today as well.

By the way, the new "inclusion" logic seems to solve #2206.

Reply:
Yes, you're correct @tsaarni. I'll try and update this to be closer to the spec.

Also, yes, I know that #2206 is partially solved by HTTPRoutes, although they currently cannot be included in themselves (so you can only do the equivalent of one layer of HTTPProxy includes). There was a vision of doing that initially for the service-apis, but I'm not sure where it's at, and it would add significant complexity to processing of the HTTPRoute objects.